### PR TITLE
fix: gateway use cf dns to prevent rate limit instead of durable object

### LINKED ISSUE #17: Replace gateway Durable Object for Rate Limit Prevention
@vasco-santos commented on [Thu Mar 17 2022](https://github.com/nftstorage/nft.storage/issues/1652)

Currently Rate Limit Prevention Durable Object is a bottleneck that might cause problems as gateway scales with errors like https://developers.cloudflare.com/workers/learning/using-durable-objects/#error-durable-object-is-overloaded

While we should definitely handle these errors for other Durable Objects, they are not so critical given they are only used after a response is provided to the user (but we won't collect metrics). However, if the Rate Limit Prevention DO is overwhelmed, users are affected.

There are a few ways we can move around this:

## Do not perform Rate Limit Prevention

We have a solution that works within Durable Object limits and is good. Its main motivation was when we were being aggressively rate limited by Public Gateways. Since then, we have solutions in place to be rate limited by user IP instead of worker IP for ipfs.io and Cloudflare gateways.
Given dedicated Pinata gateways do not have rate limits mechanisms in place, we could consider just not performing rate limit prevention at all.
The main disadvantage of this option is that we might overwhelm Pinata and force them to add Rate Limit (and of course hit their performance to handle requests), as well as adding other gateways that also have rate limits will become a problem.

## Rely on CNAME DNS Record and CF Firewall Rate Limits

> This solution would build on top of "Do not perform Rate Limit Prevention"

We can also remove Rate Limit Prevention from worker code, and push that responsibility within Cloudflare Firewall. For this, we would simply need to setup a DNS CNAME record for each gateway we want to support like `pinata.nftstorage.link` pointing to the real dedicated gateway. We can then setup Firewall to only allow requests from our worker to it and setup the rate limits we want to use there.
The main disadvantage for this solution is that we need to do this configuration for each new gateway we add. However, this might not be so bad as we might want different rate limits per gateway.

## Rely on a new worker for generic rate limiting by query parameter

> This solution would build on top of "Do not perform Rate Limit Prevention"

We can have a worker that just proxies requests to public gateways and rate limit using a query parameter. This would allow us to use a DNS record like `gw.nftstorage.link?gw='pinata'`. This solution would help us not needing to configure new Firewall/Rate limit rules by just updating an ENV var in the worker.

The main disadvantages are that this solution would likely be more pricey, would need code and if we would like different rate limit strategies we would need to hardcode them in the code.

Note: I think we would need to do a redirect here and not proxy traffic or we would likely need to pay for incoming traffic? So, we would need an extra hop.

## Conclusions

I think solution "Rely on CNAME DNS Record and CF Firewall Rate Limits" provides us the most flexibility and is easy to implement. We can easily just setup the CF rules as needed as they will mostly be copy paste of already created ones. They have limits of rules we can have, but it is 100 rules, so we are fine on that.

---

@olizilla commented on [Thu Mar 17 2022](https://github.com/nftstorage/nft.storage/issues/1652#issuecomment-1070812013)

Can we just put a CF Firewall rate limit on nftstorage.link directly? If we're racing gateways, we can tollerate some degree of being rate limited by others.

---

@vasco-santos commented on [Thu Mar 17 2022](https://github.com/nftstorage/nft.storage/issues/1652#issuecomment-1070891298)

> Can we just put a CF Firewall rate limit on nftstorage.link directly? If we're racing gateways, we can tollerate some degree of being rate limited by others.

@olizilla We have it already https://github.com/nftstorage/nft.storage/tree/main/packages/gateway#rate-limiting

The core problem here in terms of rate limiting is that typically gateways will rate limit by IP and therefore they will rate limit nftstorage.link rather than the client IP and the blockage can be super long. This way, we have been doing some work to work around this as you can read history on https://github.com/nftstorage/nft.storage/issues/1195

We started by creating the concept of Rate Limit Prevention Durable Object, which kept state of timestamp from requests within the timeframe of their rate limit properties. This allowed us to prevent requests that would rate limit our worker from performing requests. For you to have a better understanding of the impact here, CF public gateway would block us completely for 60 minutes if we got rate limited.

From that initial point, we moved with ipfs.io and CF teams with better solutions to work around this. Namely, ipfs.io now supports `X-Forwarded-For` header to understand we are proxying and rate limiting the client. I had some talks with CF and they kindly provided us a beta feature with a dedicated gateway manually setup via a DNS record from our own CF zone. This means that we now have `cf.nftstorage.link` pointing to this dedicated gateway and we can tweak its configuration of rate limiting instead of being under the strict rate limits in the public gateway. For better optimization of resources, we have a firewall in place there that does not allow any request to `cf.nftstorage.link` outside the `nftstorage.link` CF zone.

The proposals 2 and 3 here, are to do something similar to what we do with CF dedicated gateway in CF Firewall/rate limit setup, rather than relying on tracking state in the Durable Object

---

@olizilla commented on [Thu Mar 17 2022](https://github.com/nftstorage/nft.storage/issues/1652#issuecomment-1070944900)

Thanks for the context! I prefer
> Rely on CNAME DNS Record and CF Firewall Rate Limits

### PR DESCRIPTION
**Pull request by vasco-santos**
Per Cloudflare durable limits that lead to temporary fix on https://github.com/nftstorage/nftstorage.link/pull/32 , this PR adds solution proposed in #17 by replacing Durable Object logic to prevent rate limits by a DNS based solution that enables us to rely on CF dashboard to stipulate rate limits.

Added rate limit rule https://dash.cloudflare.com/fffa4b4363a7e5250af8357087263b3a/nftstorage.link/security/waf/rate-limiting-rules `pinata.nftstorage.link by X-Forwarded-For`. Internally within the worker, we track 429 errors already and wey will be counted in metrics.

TODO:
- [x] Add domain to CF nftstorage.link zone
- [x] Setup rate limits to pinata.nftstorage.link

Closes #17

### CODE REVIEW

**Comment by vasco-santos** [comment]
Note: there is currently an error "CNAME cross-user banned". This error happens when we attempt to CNAME to a Cloudflare proxied site. We will need to talk with Pinata to have our domain whitelisted

**Comment by vasco-santos** [comment]
> Note: there is currently an error "CNAME cross-user banned". This error happens when we attempt to CNAME to a Cloudflare proxied site. We will need to talk with Pinata to have our domain whitelisted

This was arranged with Pinata team and we are now live 🎉  https://pinata.nftstorage.link/ipfs/bafybeiegsdpbk5bs73mtsj2cduyaq3hddra5c33t6vw2vcctxhhxmsd46u

File: packages/edge-gateway/wrangler.toml [hunk 1]
```diff
@@ -148,4 +142,8 @@ deleted_classes = ["GatewayRateLimits2"]
 [[migrations]]
 tag = "v5" # Should be unique for each entry
 new_classes = ["GatewayRateLimits4"]
-deleted_classes = ["GatewayRateLimits3"]
\ No newline at end of file
+deleted_classes = ["GatewayRateLimits3"]
+
+[[migrations]]
+tag = "v5" # Should be unique for each entry
```
Comment (olizilla) [inline packages/edge-gateway/wrangler.toml:148]: Should be unique for each entry!
```suggestion
tag = "v6" # Should be unique for each entry
```

**Review by olizilla — CHANGES_REQUESTED** [review]

**Review by alanshaw — APPROVED** [review]
LGTM

**Review by olizilla — APPROVED** [review]

**Pull request merged**